cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -529,7 +529,7 @@ class KeyHasher {
   size_t index_;
   std::vector<col_index_t> indices_;
   std::vector<KeyColumnMetadata> metadata_;
-  const RecordBatch* batch_;
+  std::atomic<const RecordBatch*> batch_;
```

Comment:
Above it is saying "the key hasher is not thread-safe", if so, why do we care if this is atomic here?

Reply:
Are we using the key hasher in multiple thread?

Reply:
It is queried from one thread but invalidated from another. We discussed offline that this can be simplified so that the key hasher would only be used from one thread, but this is (currently?) out of scope for this PR.

Reply:
I see in that case you can update the doc to explain how this class should be used in multi-threaded execution? Currently looks like the doc about thread safety is not correct, i.e., it is used by multiple thread while the docstring says it's not thread safe.

Reply:
I ended up just fixing it to be single-threaded.

Reply:
> I ended up just fixing it to be single-threaded.
Nice - can you point me where the fix is?

Reply:
Sorry, I only pushed it in [the recent commit](https://github.com/apache/arrow/pull/35874/commits/e29520b4974e9bbc4ef08687913ce0199ab08b8a).

Reply:
@rtpsw Do we still need to change this. If so, why?